### PR TITLE
Multipart [WIP]

### PR DESCRIPTION
**Pull request by ChristopherDavenport**
Here is an initial implementation that works for Encoder and Decoder and has tests working again. Here are the steps I see that I'd like to outline

- [x] MultipartEncoder
- [x] MultipartDecoder
- [x] Tests
- [ ] Parser
    - [ ] Implement Pulls with Handle Passing
    - [ ] Parse Headers
    - [ ] Parse Body(including Binary, etc)

### CODE REVIEW

File: core/src/main/scala/org/http4s/multipart/MultipartEncoder.scala [hunk 1]
```diff
@@ -1,66 +1,74 @@
...
-      } << Boundary.CRLF).toByteVector) ++ p.body
-
-    val parts = mp.parts
-    val body = parts.tail.foldLeft(renderPart(_start, parts.head)) { (acc, part) =>
-      acc ++ renderPart(_encapsulation, part)
-    } ++ Process.emit(_end)
-
-    Task.now(Entity(body, None))
-  }
+  def toEntity(mp: Multipart): Task[Entity] = Task.now(Entity(renderParts(mp.boundary)(mp.parts), None))
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/multipart/MultipartEncoder.scala]: I am glad to see this odd formatting getting cleaned up.

File: core/src/main/scala/org/http4s/multipart/MultipartParser.scala [hunk 2]
```diff
@@ -1,236 +1,169 @@
...
 
-  private val CRLF = ByteVector('\r', '\n')
-  private val DASHDASH = ByteVector('-', '-')
+  private val CRLF = "\r\n"
+  private val DASHDASH = "--"
+  private val startLine: Boundary => String = boundary => s"$DASHDASH${boundary.value}"
+  private val endLine: Boundary => String = boundary => s"${startLine(boundary)}$DASHDASH"
+  private val expected: Boundary => String = boundary => s"$CRLF${startLine(boundary)}"
+
+  private val CRLFBytes = ByteVector('\r','\n')
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/multipart/MultipartParser.scala]: How do we end up with ByteVector and String versions of these?

File: core/src/main/scala/org/http4s/multipart/MultipartParser.scala [hunk 3]
```diff
@@ -1,236 +1,169 @@
...
+
+    val streamString = s.chunks
+      .through(text.utf8DecodeC)
+      .covary[Task]
+      .through(text.lines)
+      .dropWhile(_ != startLine(boundary)) // Drop Prelude
+      .drop(1)                             // Drop Start Line
+      .takeWhile(_ != endLine(boundary))   // Take Until EndLine
+
+      val headers =
```
Comment (rossabaker) [inline core/src/main/scala/org/http4s/multipart/MultipartParser.scala]: Indentation is weird here.

**Comment by ChristopherDavenport** [comment]
Update: I think fastest way to get this up is to buffer and then parse. I have quite a solid async parsing program that I like but it make take a bit. Pushing into queues, and then consuming them within the method. 

Would doing a large buffer to get this operational faster be an ok middle implementation, and then working on a more lasting streaming version as a second wave.

**Comment by rossabaker** [comment]
:+1: I'd rather have a suboptimal implementation than no implementation.

**Comment by rossabaker** [comment]
Is this superseded by #1238 ?

**Comment by ChristopherDavenport** [comment]
I would say so, the ideas were good here, but buffered will get us and running and this PR, does not aid that anymore.

**Pull request closed without merging**